src/databricks/labs/blueprint/paths.py

```diff
@@ -448,17 +441,296 @@ def match(self, path_pattern, *, case_sensitive=None):
+            return True
+        except NotFound:
+            return False
+
+    def mkdir(self, mode: int = 0o600, parents: bool = True, exist_ok: bool = True) -> None:
+        """Create a directory in DBFS. Only mode 0o600 is supported."""
+        if not exist_ok:
+            raise ValueError("exist_ok must be True for Databricks Workspace")
+        if not parents:
+            raise ValueError("parents must be True for Databricks Workspace")
+        if mode != 0o600:
+            raise ValueError("other modes than 0o600 are not yet supported")
+        self._ws.dbfs.mkdirs(self.as_posix())
+
+    def rmdir(self, recursive: bool = False) -> None:
+        """Remove a DBFS directory"""
+        self._ws.dbfs.delete(self.as_posix(), recursive=recursive)
+
+    def _rename(self: P, target: str | bytes | os.PathLike, overwrite: bool) -> P:
+        dst = self.with_segments(target)
+        if overwrite:
+            with dst.open(mode="wb") as writer, self.open(mode="rb") as reader:
+                shutil.copyfileobj(reader, writer, length=1024 * 1024)
+            self.unlink()
+        else:
+            self._ws.dbfs.move(self.as_posix(), dst.as_posix())
+        return dst
+
+    def unlink(self, missing_ok: bool = False) -> None:
+        """Remove a file in Databricks Workspace."""
+        if not missing_ok and not self.exists():
+            raise FileNotFoundError(f"{self.as_posix()} does not exist")
+        self._ws.dbfs.delete(self.as_posix())
+
+    def open(
+        self,
+        mode: str = "r",
+        buffering: int = -1,
+        encoding: str | None = None,
+        errors: str | None = None,
+        newline: str | None = None,
+    ):
+        """Open a DBFS file.
+
+        Only text and binary I/O are supported in basic read or write mode, along with 'x' to avoid overwriting."""
+        is_write = "w" in mode
+        is_read = "r" in mode or not is_write
+        if is_read and is_write:
+            msg = f"Unsupported mode: {mode} (simultaneous read and write)"
+            raise ValueError(msg)
+        is_binary = "b" in mode
+        is_text = "t" in mode or not is_binary
+        if is_binary and is_text:
+            msg = f"Unsupported mode: {mode} (binary and text I/O)"
+            raise ValueError(msg)
+        is_overwrite = is_write and "x" not in mode
+        binary_io = self._ws.dbfs.open(self.as_posix(), read=is_read, write=is_write, overwrite=is_overwrite)
+        if is_text:
+            return io.TextIOWrapper(binary_io, encoding=encoding, errors=errors, newline=newline)
+        return binary_io
+
+    def write_bytes(self, data):
+        """Write the (binary) data to this path."""
+        # The DBFS BinaryIO implementation only accepts bytes and rejects the (other) byte-like builtins.
+        match data:
+            case builtins.bytes | builtins.bytearray:
+                binary_data = bytes(data)
+            case _:
+                binary_data = bytes(memoryview(data))
```

Comment:
example of this?

Reply:
Python's own `array.array()` supports it, as does `numpy`: it's the Python way of exposing the underlying memory buffer directly for I/O.

This method was implemented because the default `.write_bytes()` implementation wraps everything as a `memoryview` before trying to write to the file, but the DBFS file-like object it gets from `.open()` only accepts `bytes`. I therefore introduced this implementation to accept the normal types allowed by the method contract while supplying DBFS with bytes.

One thing worth mentioning: there's no match arm for `bytes` on its own because assigning directly confuses Mypy which doesn't recognise the instance check, and `bytes(b)` is a no-op if it's already `bytes`.